README.md
# FCWC Data Processing

The code in this repository are written in the [R language](https://cran.r-project.org) and can be used to process data collected by an [In-Situ Aqua TROLL 600 Multiparameter sonde](https://in-situ.com/us/aqua-troll-600-multiparameter-sonde). The data are collected by the non-profit Florida Commercial Watermen's Conservation ([FCWC](https://floridawatermen.org)) and is a group of commercial fishermen located in southwest Florida. The data are transmitted to scientists at NOAA's Southeast Fisheries Science Center ([SEFSC](https://www.fisheries.noaa.gov/about/southeast-fisheries-science-center)), NOAA's Atlantic Oceanographic and Meteorological Laboratory ([AOML](https://www.aoml.noaa.gov/)), and University of Miami's Cooperative Institute for Marine and Atmospheric Studies ([CIMAS](https://cimas.rsmas.miami.edu/)) for processing and distribution. The data undergo quality assurance and quality control (QA/QC) and then visualized products are sent back to the collaborating fishermen and other stakeholders. The data products are a work in progress with the goal of being responsive to the needs of the fishermen and to produce informative data products that can be used for planning purposes.

The original files are directly uploaded to GoogleDrive by the [VuSitu App](https://in-situ.com/us/vusitu-app) as an htm. The R functions in this repository are designed to perform a few different tasks: 

* **aquatroll_fields** a helper function that just lists the standard parameters reported in aquatroll htm files
* **summary_aquatroll** takes an .htm or .csv and returns a summary data.frame of the data including if parameters were missing from the input file or if there was sensor drop out, both of which has been an issue
* **data_extract_aquatroll** takes an .htm or .csv and extract all the raw data and put it into a data.frame
* **qaqc_automation** performs quality assurance and quality control based upon Integrated Ocean Observing System ([IOOS](https://ioos.noaa.gov/project/qartod/)) recommendations
* **interp_aquatroll** takes the data.frame output of **data_extract_aquatroll** and smooths and interpolates selected parameters for downstream visualization
* **bottom_finder** is useful to find bottom for plotting profiles
* **breaks** makes handy breaks for color palettes
* **running** calculates a moving linear regression slope and moving average

Code to automate QA/QC is under development using guidelines outlined by the IOOS Quality Assurance / Quality Control of Real Time Oceanographic Data ([QARTOD](https://ioos.noaa.gov/project/qartod/)). The **summary_aquatroll** function will return missing parameters from a list of expected parameters and a list of parameters that dropped out. Drop out is when the parameter has some measurements but not for the whole time the unit was recording. This can happen when there is a bad connection or other sensor failure. If this happens, it is recommended that In-Situ technical support be contacted for further help.


---

# Details

The overall objective of the code is produce a visual data product that can be distributed to various stakeholders, including the fishermen that collect the data.

![alt text](https://github.com/imaginaryfish/FCWC-data-processing/blob/master/figures/FCWC_2019-Oct-12.png "Example data product")

---

# Disclaimer

*This repository is a scientific product and is not official communication of the National Oceanic and Atmospheric Administration, or the United States Department of Commerce. All NOAA GitHub project code is provided on an ‘as is’ basis and the user assumes responsibility for its use. Any claims against the Department of Commerce or Department of Commerce bureaus stemming from the use of this GitHub project will be governed by all applicable Federal law. Any reference to specific commercial products, processes, or services by service mark, trademark, manufacturer, or otherwise, does not constitute or imply their endorsement, recommendation or favoring by the Department of Commerce. The Department of Commerce seal and logo, or the seal and logo of a DOC bureau, shall not be used in any manner to imply endorsement of any commercial product or activity by DOC or the United States Government.*

---

To do 

* add QA/QC processing functions (in progess)
* make shiny app to display data https://shiny.rstudio.com/
* update .md using: https://r4ds.had.co.nz/r-markdown-formats.html or https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet or https://rmarkdown.rstudio.com/ examples



docs/changelog.txt
2023-01-06  Brendan Turley
- added look_aquatroll function to exctract raw data without assuming which sensors are installed; also returns serial numbers for sensors

2022-11-18  Brendan Turley
- updated input data headers; In-Situ changed the salinity units; add serial number to aquatroll_interp output plots

2022-08-23  Brendan Turley
- qaqc_automation_vs.R: updating QAQC code to be more flexible

2021-10-19  Brendan Turley
- shiny_prototype.R: wokring on Shiny app to display data

2021-10-19  Brendan Turley
- aquatroll_processing.R: interp_aquatroll function updated to output legend in profile plots; summary_aquatroll function fixed Marked "coerced as NA" warning

2021-10-18  Brendan Turley
- aquatroll_processing.R: interp_aquatroll function changed to output chlorophyll fluoresence instead of concentration; changed output plot to give lat/lon at top for reference

2021-10-14  Brendan Turley
- qaqc_automation.R: added time/gap, location, gross range, climatological, spike, and rate of change tests

2021-10-13  Brendan Turley
- qaqc_automation.R: added outline for QAQC methods based upon IOOS recommendations

2021-10-06  Brendan Turley
- aquatroll_processing.R: data_extract_aquatroll includes input filename in output; interp_aquatroll plots filename for reference

2021-10-05  Brendan Turley
- aquatroll_processing.R: interp_aquatroll function changed to deal with NAs during smooth, bin, and interpolation process because smooth.spline functino does not accept NAs; future addition possible to use upcast data if downcast is bad/unacceptable

2021-10-04  Brendan Turley
- aquatroll_processing.R: remove process_aquatroll function; it was redundant with newer functions performing the same operations more efficiently; fixed chlorophyll dropout replacement in interp_aquatroll function; changed 'too few data' stop to a warning and added else statement to prevent full stop
- process_aquatroll.R: depreciated function moved to new folder

2021-10-01  Brendan Turley
- aquatroll_processing.R: added aquatroll_fields helper function that lists standard aquatroll output parameters

2021-09-30  Brendan Turley
- aquatroll_processing.R: interp_aquatroll function option save_plot

2021-09-29  Brendan Turley
- aquatroll_processing.R: interp_aquatroll now saves current working directory so it returns to htm locations after saving profile plots in different directory

2021-09-27  Brendan Turley
- aquatroll_processing.R: fix interp_aquatroll so it does not look for profile index and now includes lat/lon as output; update bottom_finder funciton so output can be used as a polygon without extra processing
- FCWC_vignette.R: updated wored example to produce interpolated profiles (with colorbars!)

2021-09-24  Brendan Turley
- aquatroll_processing.R: update data_extract_aquatroll to export numeric columns because direct pipe into interp_aquatroll was causing errors; update interp_aquatroll to no longer recognize profile index column, which I added in a processing step outside these defined functions
- added FCWC_vignette.R to show example processing pipeline; this is a work in progress

2021-09-23  Brendan Turley
- aquatroll_processing.R: update function parameter listings; update interp_aquatroll to plot interpolated data for inspection

2021-09-22  Brendan Turley
- aquatroll_processing.R: added new interpolation function interp_aquatroll to replace old function process_aquatroll

2021-09-20  Brendan Turley
- aquatroll_processing.R: removal of depreciated and redundant functions dropout_aquatroll, process_htm, summary_aquatroll, summary_htm

2021-07-30  Brendan Turley
- aquatroll_processing.R: fixed error coding characters for spltstr function; added bottom finder function and breaks function

2021-08-02  Brendan Turley
- aquatroll_processing.R: added new smoothing, depth binning, and interpolation method

2021-08-11  Brendan Turley
- aquatroll_processing.R: added conversion from depth in feet to meters; reorder columns to be consistent

2021-08-13  Brendan Turley
- aquatroll_processing.R: created new functions to process either csv or htms

R_code/qaqc_automation.R
### aquatrol quality control

### Recommended methods to comply with IOOS standards
# https://ioos.noaa.gov/project/qartod/

### flags
# 1 = Pass
# 2 = Not evaluated
# 3 = Suspect or Of High Interest
# 4 = Fail
# 5 = suspect spike test
# 6 = fail spike test
# 7 = suspect rate of change test
# 9 = Missing data

### tests
# Group 1
# Required
# Test 1) Timing/Gap Test
# Test 2) Syntax Test
# Test 3) Location Test
# Test 4) Gross Range Test - find sensor ranges https://in-situ.com/pub/media/support/documents/AquaTROLL600_Spec-Sheet.pdf
# Test 5) Climatological Test

# Group 2
# Strongly Recommended
# Test 6) Spike Test
# Test 7) Rate of Change Test
# Test 8) Flat Line Test

# Group 3
# Suggested
# Test 9) Multi-Variate Test
# Test 10) Attenuated Signal Test
# Test 11) Neighbor Test
# Test 12) TS Curve/Space Test
# Test 13) Density Inversion Test
  
### test data
library(lubridate)
library(ncdf4)
library(rgdal)

# source('~/Desktop/professional/projects/Postdoc_FL/scripts/FCWC/FCWC_process_htm.R')
source('~/Documents/R/Github/FCWC-data-processing/R_code/aquatroll_processing.R')

### find htm files
files_wd <- '~/Desktop/professional/projects/Postdoc_FL/data/FCWC/new/CaptivaPass'
setwd(files_wd)
files <- list.files()
ind <- grep('htm',files)

input1 <- files[2]
file_sum <- summary_aquatroll(input1)
input <- data_extract_aquatroll(input1)
 

### initialize unchecked flag matrix
flags <- matrix(2,dim(input)[1],dim(input)[2]-4)
###--------- Test 1) Timing/Gap Test
ind <- grep('date',names(input),ignore.case=T)
flags[,ind] <- 1
time_stamp <- ymd_hms(input[,ind])
dt <- make_difftime(diff(time_stamp),units='seconds')
n_flags <- which(abs(dt)>1)
if(length(n_flags)>0){
  flags[n_flags,ind] <- 4
}

###--------- Test 3) Location Test
ind_lon <- grep('longitude',names(input),ignore.case=T)
ind_lat <- grep('latitude',names(input),ignore.case=T)
flags[,ind_lon] <- 1
flags[,ind_lat] <- 1
lon <- input[,ind_lon]
lat <- input[,ind_lat]
### impossible locations
lon_f2 <- which(abs(lon)>180)
lat_f2 <- which(abs(lat)>90)
if(length(lon_f2)>0){
  flags[lon_f2,ind_lon] <- 4
}
if(length(lat_f2)>0){
  flags[lat_f2,ind_lat] <- 4
}

###--------- Test 4) Gross Range Test
### these values are from the Aquatroll specification manual: https://in-situ.com/us/aqua-troll-600-multiparameter-sonde
###--------- Test 5) Climatological Test
### these values are from World Ocean Database 2018 User's Manual ver 0.10; appendix 11
###--------- Test 6) Spike Test
###--------- Test 7) Rate of Change Test
# temperature
ind <- grep('temperature',names(input),ignore.case=T)
flags[,ind] <- 1
temp <- input[,ind]
### out of range
t1_f2 <- which(temp[,1]>50 | temp[,1]<(-5))
t2_f2 <- which(temp[,2]>50 | temp[,2]<(-5))
if(length(t1_f2)>0){
  flags[t1_f2,ind[1]] <- 4
}
if(length(t2_f2)>0){
  flags[t2_f2,ind[2]] <- 4
}
### climatology only for water temperature
ind_dep <- grep('depth',names(input),ignore.case=T)
z <- input[,ind_dep[1]] ### depth used elsewhere for climatology
t1_f3 <- which(z<100 & temp[,1]>35 | z<100 & temp[,1]<(-2.10))
t1_f4 <- which(z>=100 & temp[,1]>30 | z>=100 & temp[,1]<(-2.10))
if(length(t1_f3)>0){
  flags[t1_f3,ind[1]] <- 3
}
if(length(t1_f4)>0){
  flags[t1_f4,ind[1]] <- 3
}
### spike test; 6 = fail; 5 = suspect
threshold_high <- 5*sd(temp[,1],na.rm=T)
threshold_low <- 3*sd(temp[,1],na.rm=T)
for(i in 2:(length(temp[,1])-1)){
  threshold <- mean(temp[c((i-1),(i+1)),1],na.rm=T)
  if(abs(temp[i,1]-threshold)>threshold_high &
     flags[i,ind[1]]==1){
    flags[i,ind[1]] <- 6
  } else if (abs(temp[i,1]-threshold)>threshold_low &
             abs(temp[i,1]-threshold)<=threshold_high &
             flags[i,ind[1]]==1){
    flags[i,ind[1]] <- 5
  }
}
### rate of change test; 7 = suspect
threshold_low <- 3*sd(temp[,1],na.rm=T)
for(i in 1:(length(temp[,1])-1)){
  if(abs(temp[i,1]-temp[(i+1),1])>threshold_low &
     flags[i,ind[1]]==1){
    flags[i,ind[1]] <- 7
  }
}
# conductivity
ind <- grep('conductivity',names(input),ignore.case=T)
flags[,ind] <- 1
cond <- input[,ind]
### out of range
c1_f2 <- which(cond[,1]>350000 | cond[,1]<0)
c2_f2 <- which(cond[,2]>350000 | cond[,2]<0)
if(length(c1_f2)>0){
  flags[c1_f2,ind[1]] <- 4
}
if(length(c2_f2)>0){
  flags[c2_f2,ind[2]] <- 4
}
# salinity
ind <- grep('salinity',names(input),ignore.case=T)
flags[,ind] <- 1
sal <- input[,ind]
### out of range
s_f2 <- which(sal>350 | sal<0)
if(length(s_f2)>0){
  flags[s_f2,ind] <- 4
}
### climatology
s_f3 <- which(z<50 & sal>40 | z<50 & sal<0)
s_f4 <- which(z>=50 & sal>40 | z>=50 & sal<20)
if(length(s_f3)>0){
  flags[s_f3,ind] <- 3
}
if(length(s_f4)>0){
  flags[s_f4,ind] <- 3
}
### spike test; 6 = fail; 5 = suspect
threshold_high <- 5*sd(sal,na.rm=T)
threshold_low <- 3*sd(sal,na.rm=T)
for(i in 2:(length(sal)-1)){
  threshold <- mean(sal[c((i-1),(i+1))],na.rm=T)
  if(abs(sal[i]-threshold)>threshold_high &
     flags[i,ind]==1){
    flags[i,ind] <- 6
  } else if (abs(sal[i]-threshold)>threshold_low &
             abs(sal[i]-threshold)<=threshold_high &
             flags[i,ind]==1){
    flags[i,ind] <- 5
  }
}
### rate of change test; 7 = suspect
threshold_low <- 3*sd(sal,na.rm=T)
for(i in 1:(length(sal)-1)){
  if(abs(sal[i]-sal[(i+1)])>threshold_low &
     flags[i,ind]==1){
    flags[i,ind] <- 7
  }
}
# chlorophyll
ind <- grep('chlorophyll',names(input),ignore.case=T)
flags[,ind] <- 1
chl <- input[,ind]
### out of range
c1_f2 <- which(chl[,1]>100 | chl[,1]<0)
c2_f2 <- which(chl[,2]>1000 | chl[,2]<0)
if(length(c1_f2)>0){
  flags[c1_f2,ind[1]] <- 4
}
if(length(c2_f2)>0){
  flags[c2_f2,ind[2]] <- 4
}
### spike test; 6 = fail; 5 = suspect
threshold_high <- 5*sd(chl[,1],na.rm=T)
threshold_low <- 3*sd(chl[,1],na.rm=T)
for(i in 2:(length(chl[,1])-1)){
  threshold <- mean(chl[c((i-1),(i+1)),1],na.rm=T)
  if(abs(chl[i,1]-threshold)>threshold_high &
     flags[i,ind[1]]==1){
    flags[i,ind[1]] <- 6
  } else if (abs(chl[i,1]-threshold)>threshold_low &
             abs(chl[i,1]-threshold)<=threshold_high &
             flags[i,ind[1]]==1){
    flags[i,ind[1]] <- 5
  }
}
threshold_high <- 5*sd(chl[,2],na.rm=T)
threshold_low <- 3*sd(chl[,2],na.rm=T)
for(i in 2:(length(chl[,2])-1)){
  threshold <- mean(chl[c((i-1),(i+1)),2],na.rm=T)
  if(abs(chl[i,2]-threshold)>threshold_high &
     flags[i,ind[1]]==1){
    flags[i,ind[1]] <- 6
  } else if (abs(chl[i,2]-threshold)>threshold_low &
             abs(chl[i,2]-threshold)<=threshold_high &
             flags[i,ind[1]]==1){
    flags[i,ind[1]] <- 5
  }
}
### rate of change test; 7 = suspect
threshold_low <- 3*sd(chl[,1],na.rm=T)
for(i in 1:(length(chl[,1])-1)){
  if(abs(chl[i,1]-chl[(i+1),1])>threshold_low &
     flags[i,ind[1]]==1){
    flags[i,ind[1]] <- 7
  }
}
threshold_low <- 3*sd(chl[,2],na.rm=T)
for(i in 1:(length(chl[,2])-1)){
  if(abs(chl[i,2]-chl[(i+1),2])>threshold_low &
     flags[i,ind[2]]==1){
    flags[i,ind[2]] <- 7
  }
}
# oxygen
ind <- grep('rdo concentration',names(input),ignore.case=T)
flags[,ind] <- 1
rdo <- input[,ind]
### out of range
o_f2 <- which(rdo>50 | rdo<0)
if(length(o_f2)>0){
  flags[o_f2,ind] <- 4
}
### climatology
o_f3 <- which(z<50 & rdo>436*1.025/31.26 | z<50 & rdo<0)
o_f4 <- which(z>=50 & rdo>392*1.025/31.26 | z>=50 & rdo<0)
if(length(o_f3)>0){
  flags[o_f3,ind] <- 3
}
if(length(o_f4)>0){
  flags[o_f4,ind] <- 3
}
### spike test; 6 = fail; 5 = suspect
threshold_high <- 5*sd(rdo,na.rm=T)
threshold_low <- 3*sd(rdo,na.rm=T)
for(i in 2:(length(rdo)-1)){
  threshold <- mean(rdo[c((i-1),(i+1))],na.rm=T)
  if(abs(rdo[i]-threshold)>threshold_high &
     flags[i,ind]==1){
    flags[i,ind] <- 6
  } else if (abs(rdo[i]-threshold)>threshold_low &
             abs(rdo[i]-threshold)<=threshold_high &
             flags[i,ind]==1){
    flags[i,ind] <- 5
  }
}
### rate of change test; 7 = suspect
threshold_low <- 3*sd(rdo,na.rm=T)
for(i in 1:(length(rdo)-1)){
  if(abs(rdo[i]-rdo[(i+1)])>threshold_low &
     flags[i,ind]==1){
    flags[i,ind] <- 7
  }
}
### global missing data test; if NA
na_flag <- which(is.na(input))
if(length(na_flag)>0){
  flags[na_flag] <- 9
}
### column names
flags <- data.frame(flags)
names(flags) <- names(input)[1:(ncol(input)-4)]

###--------- aquatroll quality control
qaqc_aquatroll <- function (input # input file is the output data.frame from data_extract_aquatroll function
)
{
  ### initialize unchecked flag matrix
  flags <- matrix(2,dim(input)[1],dim(input)[2]-4)
  
  ###--------- Test 1) Timing/Gap Test
  ind <- grep('date',names(input),ignore.case=T)
  flags[,ind] <- 1
  time_stamp <- ymd_hms(input[,ind])
  dt <- make_difftime(diff(time_stamp),units='seconds')
  n_flags <- which(abs(dt)>1)
  if(length(n_flags)>0){
    flags[n_flags,ind] <- 4
  }
  
  ###--------- Test 3) Location Test
  ind_lon <- grep('longitude',names(input),ignore.case=T)
  ind_lat <- grep('latitude',names(input),ignore.case=T)
  flags[,ind_lon] <- 1
  flags[,ind_lat] <- 1
  lon <- input[,ind_lon]
  lat <- input[,ind_lat]
  ### impossible locations
  lon_f2 <- which(abs(lon)>180)
  lat_f2 <- which(abs(lat)>90)
  if(length(lon_f2)>0){
    flags[lon_f2,ind_lon] <- 4
  }
  if(length(lat_f2)>0){
    flags[lat_f2,ind_lat] <- 4
  }
  
  ###--------- Test 4) Gross Range Test
  ### these values are from the Aquatroll specification manual: https://in-situ.com/us/aqua-troll-600-multiparameter-sonde
  ###--------- Test 5) Climatological Test
  ### these values are from World Ocean Database 2018 User's Manual ver 0.10; appendix 11
  ###--------- Test 6) Spike Test
  ###--------- Test 7) Rate of Change Test
  # temperature
  ind <- grep('temperature',names(input),ignore.case=T)
  flags[,ind] <- 1
  temp <- input[,ind]
  ### out of range
  t1_f2 <- which(temp[,1]>50 | temp[,1]<(-5))
  t2_f2 <- which(temp[,2]>50 | temp[,2]<(-5))
  if(length(t1_f2)>0){
    flags[t1_f2,ind[1]] <- 4
  }
  if(length(t2_f2)>0){
    flags[t2_f2,ind[2]] <- 4
  }
  ### climatology only for water temperature
  ind_dep <- grep('depth',names(input),ignore.case=T)
  z <- input[,ind_dep[1]] ### depth used elsewhere for climatology
  t1_f3 <- which(z<100 & temp[,1]>35 | z<100 & temp[,1]<(-2.10))
  t1_f4 <- which(z>=100 & temp[,1]>30 | z>=100 & temp[,1]<(-2.10))
  if(length(t1_f3)>0){
    flags[t1_f3,ind[1]] <- 3
  }
  if(length(t1_f4)>0){
    flags[t1_f4,ind[1]] <- 3
  }
  ### spike test; 6 = fail; 5 = suspect
  threshold_high <- 5*sd(temp[,1],na.rm=T)
  threshold_low <- 3*sd(temp[,1],na.rm=T)
  for(i in 2:(length(temp[,1])-1)){
    threshold <- mean(temp[c((i-1),(i+1)),1],na.rm=T)
    if(abs(temp[i,1]-threshold)>threshold_high &
       flags[i,ind[1]]==1){
      flags[i,ind[1]] <- 6
    } else if (abs(temp[i,1]-threshold)>threshold_low &
               abs(temp[i,1]-threshold)<=threshold_high &
               flags[i,ind[1]]==1){
      flags[i,ind[1]] <- 5
    }
  }
  ### rate of change test; 7 = suspect
  threshold_low <- 3*sd(temp[,1],na.rm=T)
  for(i in 1:(length(temp[,1])-1)){
    if(abs(temp[i,1]-temp[(i+1),1])>threshold_low &
       flags[i,ind[1]]==1){
      flags[i,ind[1]] <- 7
    }
  }
  # conductivity
  ind <- grep('conductivity',names(input),ignore.case=T)
  flags[,ind] <- 1
  cond <- input[,ind]
  ### out of range
  c1_f2 <- which(cond[,1]>350000 | cond[,1]<0)
  c2_f2 <- which(cond[,2]>350000 | cond[,2]<0)
  if(length(c1_f2)>0){
    flags[c1_f2,ind[1]] <- 4
  }
  if(length(c2_f2)>0){
    flags[c2_f2,ind[2]] <- 4
  }
  # salinity
  ind <- grep('salinity',names(input),ignore.case=T)
  flags[,ind] <- 1
  sal <- input[,ind]
  ### out of range
  s_f2 <- which(sal>350 | sal<0)
  if(length(s_f2)>0){
    flags[s_f2,ind] <- 4
  }
  ### climatology
  s_f3 <- which(z<50 & sal>40 | z<50 & sal<0)
  s_f4 <- which(z>=50 & sal>40 | z>=50 & sal<20)
  if(length(s_f3)>0){
    flags[s_f3,ind] <- 3
  }
  if(length(s_f4)>0){
    flags[s_f4,ind] <- 3
  }
  ### spike test; 6 = fail; 5 = suspect
  threshold_high <- 5*sd(sal,na.rm=T)
  threshold_low <- 3*sd(sal,na.rm=T)
  for(i in 2:(length(sal)-1)){
    threshold <- mean(sal[c((i-1),(i+1))],na.rm=T)
    if(abs(sal[i]-threshold)>threshold_high &
       flags[i,ind]==1){
      flags[i,ind] <- 6
    } else if (abs(sal[i]-threshold)>threshold_low &
               abs(sal[i]-threshold)<=threshold_high &
               flags[i,ind]==1){
      flags[i,ind] <- 5
    }
  }
  ### rate of change test; 7 = suspect
  threshold_low <- 3*sd(sal,na.rm=T)
  for(i in 1:(length(sal)-1)){
    if(abs(sal[i]-sal[(i+1)])>threshold_low &
       flags[i,ind]==1){
      flags[i,ind] <- 7
    }
  }
  # chlorophyll
  ind <- grep('chlorophyll',names(input),ignore.case=T)
  flags[,ind] <- 1
  chl <- input[,ind]
  ### out of range
  c1_f2 <- which(chl[,1]>100 | chl[,1]<0)
  c2_f2 <- which(chl[,2]>1000 | chl[,2]<0)
  if(length(c1_f2)>0){
    flags[c1_f2,ind[1]] <- 4
  }
  if(length(c2_f2)>0){
    flags[c2_f2,ind[2]] <- 4
  }
  ### spike test; 6 = fail; 5 = suspect
  threshold_high <- 5*sd(chl[,1],na.rm=T)
  threshold_low <- 3*sd(chl[,1],na.rm=T)
  for(i in 2:(length(chl[,1])-1)){
    threshold <- mean(chl[c((i-1),(i+1)),1],na.rm=T)
    if(abs(chl[i,1]-threshold)>threshold_high &
       flags[i,ind[1]]==1){
      flags[i,ind[1]] <- 6
    } else if (abs(chl[i,1]-threshold)>threshold_low &
               abs(chl[i,1]-threshold)<=threshold_high &
               flags[i,ind[1]]==1){
      flags[i,ind[1]] <- 5
    }
  }
  threshold_high <- 5*sd(chl[,2],na.rm=T)
  threshold_low <- 3*sd(chl[,2],na.rm=T)
  for(i in 2:(length(chl[,2])-1)){
    threshold <- mean(chl[c((i-1),(i+1)),2],na.rm=T)
    if(abs(chl[i,2]-threshold)>threshold_high &
       flags[i,ind[1]]==1){
      flags[i,ind[1]] <- 6
    } else if (abs(chl[i,2]-threshold)>threshold_low &
               abs(chl[i,2]-threshold)<=threshold_high &
               flags[i,ind[1]]==1){
      flags[i,ind[1]] <- 5
    }
  }
  ### rate of change test; 7 = suspect
  threshold_low <- 3*sd(chl[,1],na.rm=T)
  for(i in 1:(length(chl[,1])-1)){
    if(abs(chl[i,1]-chl[(i+1),1])>threshold_low &
       flags[i,ind[1]]==1){
      flags[i,ind[1]] <- 7
    }
  }
  threshold_low <- 3*sd(chl[,2],na.rm=T)
  for(i in 1:(length(chl[,2])-1)){
    if(abs(chl[i,2]-chl[(i+1),2])>threshold_low &
       flags[i,ind[2]]==1){
      flags[i,ind[2]] <- 7
    }
  }
  # oxygen
  ind <- grep('rdo concentration',names(input),ignore.case=T)
  flags[,ind] <- 1
  rdo <- input[,ind]
  ### out of range
  o_f2 <- which(rdo>50 | rdo<0)
  if(length(o_f2)>0){
    flags[o_f2,ind] <- 4
  }
  ### climatology
  o_f3 <- which(z<50 & rdo>436*1.025/31.26 | z<50 & rdo<0)
  o_f4 <- which(z>=50 & rdo>392*1.025/31.26 | z>=50 & rdo<0)
  if(length(o_f3)>0){
    flags[o_f3,ind] <- 3
  }
  if(length(o_f4)>0){
    flags[o_f4,ind] <- 3
  }
  ### spike test; 6 = fail; 5 = suspect
  threshold_high <- 5*sd(rdo,na.rm=T)
  threshold_low <- 3*sd(rdo,na.rm=T)
  for(i in 2:(length(rdo)-1)){
    threshold <- mean(rdo[c((i-1),(i+1))],na.rm=T)
    if(abs(rdo[i]-threshold)>threshold_high &
       flags[i,ind]==1){
      flags[i,ind] <- 6
    } else if (abs(rdo[i]-threshold)>threshold_low &
               abs(rdo[i]-threshold)<=threshold_high &
               flags[i,ind]==1){
      flags[i,ind] <- 5
    }
  }
  ### rate of change test; 7 = suspect
  threshold_low <- 3*sd(rdo,na.rm=T)
  for(i in 1:(length(rdo)-1)){
    if(abs(rdo[i]-rdo[(i+1)])>threshold_low &
       flags[i,ind]==1){
      flags[i,ind] <- 7
    }
  }
  ### output
  ### global missing data test; if NA
  na_flag <- which(is.na(input))
  if(length(na_flag)>0){
    flags[na_flag] <- 9
  }
  ### column names
  flags <- data.frame(flags)
  names(flags) <- names(input)[1:(ncol(input)-4)]
  return(flags)
}

### test
qaqc_test <- qaqc_aquatroll(input)
hist(as.matrix(qaqc_test),breaks=seq(.5,9.5,1))
